### DISCUSSION TITLE
Pass request to the meta function

### DISCUSSION THREAD

**Discussion by Apsysikal** [remix-run/remix#8445] [Proposals]
This problem showed up when needing to generate Open Graph tags like `og:image` or `og:url`. These take a url in the form of `http(s)://host/pathname`. As these are generated in the `meta` function it would be handy to have access to the request, so we could parse out the hostname. Currently we only have the `pathname` from `matches` or `location`.

The current workaround for me and others (https://discord.com/channels/770287896669978684/1193584644994629702) is to parse the parameters out of the request in the root loader and then access it via the `matches[0].data`.

**Comment by xHomu** [comment]
Is there a reason why we don't get the request object in meta? It's always a bit odd to me, how different the props are between all the Remix server exports.

**Reply by sergiodxa** [reply]
Because MetaFunction runs client side too, so it can't access the request your server received.

**Comment by sergiodxa** [comment]
MetaFunction runs client side, so it can't have access to the request.

Instead you should return from the loader whatever your MetaFunction needs, eg return request.url or even better the og:image value.

**Reply by xHomu** [reply]
Which exports are server side? Only `action`, `loader`, `headers`? What about exports like `handle`?

**Reply by sergiodxa** [reply]
Yea only those three, handle is also sent to the browser because you can access it with useMatches, similarly because Meta component needs the MetaFunction and Links component needs LinkFunction that's why those run client side and server side

**Reply by Apsysikal** [reply]
Thank you for clarifying. I guess it's not an easy fix 😅
It just felt weird to send the host to the client as the client knew it to make a connection.

**Comment by meglio** [comment]
So, how do I read some content from a `config.ini` file and pass it through to `meta` if it is not reading any loader data when rendering root ErrorBoundary?

**Reply by sergiodxa** [reply]
You need to read it in the loader and handle the case when loader data is undefined because it failed.

**Reply by meglio** [reply]
My root loader is not failing, it is a 404 page, and thus the loader data is null for some reason that I can't understand.

**Reply by Apsysikal** [reply]
If your root loader looks something like this:
```typescript
export const loader = async ({ request }: LoaderFunctionArgs) => {
  const iniData = await readIniContent();
  const domainUrl = getDomainUrl(request);
  const user = await getUserWithRole(request);
  return json({ user, domainUrl, iniData });
};
```
you can use the data in the meta function like this:
```typescript
export const meta: MetaFunction<typeof loader> = ({ data, matches }) => {
  const { user, domainUrl, iniData } = matches[0].data; // Careful, this can be null

  return [
    { title: `${iniData}` },
  ];
};
```
Maybe you could also use a splat route as a 404: https://remix.run/docs/en/main/file-conventions/routes#splat-routes

**Reply by meglio** [reply]
```
// Careful, this can be null
```

It is indeed a null, which is my point and the issue I can't find a solution for. I fetch data form a file in the root loader, but it is not available in ErrorBoundary while the root loader is not failing, it is just a 404 page.

My goal is to build a root error boundary that is fine-tuned depending on some configuration file config.ini that should be read dynamically on the server start. For all pages, in a normal mode, so far it has been working well with exposing the data through the root loader. But with the ErrorBoundary it does not work like that - the root loader data is for some reason not available in the root ErrorBoundary.

**Reply by kiliman** [reply]
What do you mean when you say it's "just a 404 page"? Are you navigating to a non-existent route, or are you throwing a 404 response?

Also, is this config data global to the entire app? Does it depend on the request, like the host header?

**Reply by meglio** [reply]
It is a non-existent route, and I'm trying to add a customization to the ErrorBoundary that depends on the application-wide `config.ini` file. This file must not be bundled (`yarn build`), it is mounted to the docker container with using docker-compose when deploying the application. Hence I need application-wide global data. Root loader data used to do the job just fine until I decided to use this global data inside ErrorBoundary.

**Reply by kiliman** [reply]
Ok, I'm almost finished with my example.

It supports global app data that is initialized at startup and is available in your loaders, route components, and `ErrorBoundary`. It even supports Promises, so if you need to fetch or use fs promises, you can.

It also supports access to request data (for example, hostname). The `handleError` export is used to marshal the request data via the `error.message` property.

It handles both server and client-side errors.

Here's a snippet of code. Once I clean up the example and types, I'll push it to GitHub.

```ts
// global-data.ts
export const init = serverOnly$(async () =>
  setAppGlobals({
    APP_NAME: "Remix App",
    NODE_ENV: process.env.NODE_ENV,
    VITE_TEST: import.meta.env.VITE_TEST,
    CLIENT_KEY: process.env.CLIENT_KEY,
    ENV_FILE: (await readEnvFile()).split("\n"),
  })
);

function readEnvFile() {
  return fs.readFile(path.resolve(process.cwd(), ".env"), "utf-8");
}
```
```ts
// entry.server.ts
export function handleError(
  error: Error | ErrorResponse,
  { request }: LoaderFunctionArgs
) {
  const url = new URL(request.url);
  if ("data" in error) {
    const e = error as ErrorResponse;
    e.data = JSON.stringify({ message: e.data, hostname: url.hostname });
  } else {
    const e = error as Error;
    e.message = JSON.stringify({ message: e.message, hostname: url.hostname });
  }
}
```
```ts
// root.tsx
export function ErrorBoundary() {
  const error = useRouteError();
  let message = "";
  let hostname = "";
  if (isRouteErrorResponse(error)) {
    message = error.data;
  } else {
    message = (error as Error).message;
  }
  if (message.startsWith("{")) {
    const d = JSON.parse(message);
    message = d.message;
    hostname = d.hostname;
  }
  return <ClientErrorBoundary error={{ message, hostname }} isServerError={true} />;
}

export function ClientErrorBoundary({
  error,
  isServerError,
}: {
  error: Error & { hostname: string };
  isServerError: boolean;
}) {
  const appGlobals = getAppGlobals();
  let { message, hostname } = error;
  if (!hostname) {
    if (typeof document !== "undefined") {
      hostname = document.location.hostname;
    }
  }

  return (
    <>
      <h1>
        {message} @ {hostname}
      </h1>
      <h2>{isServerError ? "Server Error" : "Client Error"}</h2>
      <pre>{JSON.stringify(appGlobals, null, 2)}</pre>
    </>
  );
}
```
<img width="400" alt="image" src="https://github.com/remix-run/remix/assets/47168/cf063b75-007a-4a2b-b401-e6d41dd1c2c5">

<img width="784" alt="image" src="https://github.com/remix-run/remix/assets/47168/95fb775d-260a-4383-9986-c94fb8a98c8e">

<img width="434" alt="image" src="https://github.com/remix-run/remix/assets/47168/8f579c5f-1fcb-401b-a578-4c7ff32cfec9">

**Reply by meglio** [reply]
This looks promising, thank you. Is there a way to see it as a "diff", so that it is visible what is the default template and what is the new code that made this magic happen?

**Reply by kiliman** [reply]
Yeah, I'm going to push as a series of commits so you can see what changes were made.

**Reply by meglio** [reply]
Also, Remix should support some serverside-only `export function getRuntimeGlobalPublicApplicationData()` and make it available everywhere through some hook `useRuntimeGlobalPublicApplicationData()` to avoid gymnastics like the above. Runtime global application-wide data is a typical use of data in applications.

Alternatively, since you now have `Layout`, there can be a serverside-run `export getLayoutData()`, which is then accessible through `useLayoutData()`, and from there, it can be passed down through react context (or used in components deep down the tree via the same hook).

Yet, to be fair, I still don't get it why the root loader data is not available in the ErrorBoundary. I would assume since it is root, it is called always, and unless it fails, the data it loads should be available to everyone through the hook, including root ErrorBoundary (since the root ErrorBoundary is in some sense the root page, just the "error" version of it, but it's still root, so there can still be the loader data).

**Reply by kiliman** [reply]
I finally got my example finished and deployed. Check it out here.

https://github.com/kiliman/remix-global-data

There's even a link to a deployed app.
